File: go/linkedList/linkedList.go
package main

import (
	"fmt"
	"log"
	"strconv"
)

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
}

func (ll *LinkedList) printList() {
	if ll.head == nil {
		fmt.Println("Linked list is empty")
		return
	}

	itr := ll.head
	llstring := ""

	for itr != nil {
		llstring += strconv.Itoa(itr.data) + " ---> "
		itr = itr.next
	}

	fmt.Println(llstring)
}

func (ll *LinkedList) getLength() int {
	count := 0

	itr := ll.head
	for itr != nil {
		count += 1
		itr = itr.next
	}
	return count
}

func (ll *LinkedList) insertAtBeginning(data int) {
	temp := &Node{data: data, next: ll.head}
	ll.head = temp
}

func (ll *LinkedList) insertAtEnd(data int) {
	if ll.head == nil {
		ll.head = &Node{data: data, next: nil}
		return
	}

	itr := ll.head
	for itr.next != nil {
		itr = itr.next
	}

	itr.next = &Node{data: data, next: nil}
}

func (ll *LinkedList) insertAt(index int, data int) {
	if index < 0 || index > ll.getLength() {
		log.Fatal("Invalid index")
	}

	if index == 0 {
		temp := &Node{data: data, next: ll.head}
		ll.head = temp
		return
	}

	itr := ll.head
	count := 0

	for itr != nil {
		if count == index-1 {
			temp := &Node{data: data, next: itr.next}
			itr.next = temp
			break
		}
	}
}

func (ll *LinkedList) removeAt(index int) {
	if index < 0 || index >= ll.getLength() {
		log.Fatal("Invalid index")
	}

	if index == 0 {
		ll.head = ll.head.next
		return
	}

	itr := ll.head
	count := 0

	for itr != nil {
		if count == index-1 {
			itr.next = itr.next.next
			break
		}
	}
}

func (ll *LinkedList) insertList(dataList []int) {
	ll.head = nil
	for _, data := range dataList {
		ll.insertAtEnd(data)
	}
}
func main() {
	ll := LinkedList{}
	ll.insertAtBeginning(84)
	ll.insertAtBeginning(100)
	ll.insertAtBeginning(101)
	ll.printList()
	fmt.Println(ll.getLength())
	ll.insertAtEnd(77)
	ll.printList()
	ll.insertAt(1, 22)
	ll.printList()
	ll.removeAt(1)
	ll.printList()
	ll.insertList([]int{1, 2, 3, 4, 5})
	ll.printList()

}

File: javascript/linkedList/linkedList.js
function Node(data = null, next = null) {
    this.data = data;
    this.next = next;
}

function LinkedList() {
    this.head = null;

    this.print = function () {
        if (this.head == null) {
            console.log("Linked list is empty");
            return;
        }

        let itr = this.head;
        let llstring = "";
        while (itr) {
            llstring += itr.data + " ---> "
            itr = itr.next;
        }

        console.log(llstring);
    }
    this.getLength = function () {
        let count = 0;
        let itr = this.head;
        while (itr) {
            count += 1;
            itr = itr.next;
        }
        return count;
    }
    this.insertAtBeginning = function (data) {
        let temp = new Node(data, this.head);
        this.head = temp;
    }
    this.insertAtEnd = function (data) {
        if (this.head == null) {
            this.head = new Node(data, null);
            return;
        }
        let itr = this.head;
        while (itr.next) {
            itr = itr.next;
        }
        itr.next = new Node(data, null);
    }
    this.insertAt = function (index, data) {
        if (index < 0 || index > this.getLength()) {
            console.error("Invalid index");
            return;
        }

        if (index == 0) {
            this.insertAtBeginning(data);
            return;
        }

        let count = 0;
        let itr = this.head;


        while (itr) {
            if (count == index - 1) {
                let temp = new Node(data, itr.next);
                itr.next = temp;
                break;
            }
            itr = itr.next;
            count += 1;
        }
    }

    this.removeAt = function (index) {
        if (index < 0 || index >= this.getLength()) {
            console.error("Invalid index");
            return;
        }

        if (index == 0) {
            this.head = this.head.next;
            return;
        }

        let itr = this.head;
        let count = 0;
        while (itr) {
            if (count == index - 1) {
                itr.next = itr.next.next;
                break;
            }

            itr = itr.next;
            count += 1;
        }
    }

    this.insertValues = function (dataList) {
        this.head = null;
        dataList.forEach(element => {
            this.insertAtEnd(element);
        });
    }
}



let ll = new LinkedList();
ll.insertAtBeginning(85);
ll.insertAtBeginning(32);
console.log(ll.getLength());
ll.print();
ll.insertAtEnd(100);
console.log(ll.getLength());
ll.print();

ll.insertAt(1, 21);
ll.print();
ll.removeAt(1);
ll.print();

ll.insertValuesAtEnd(["banana", "mango", "grapes", "orange"])
ll.print();